File: interpreter_node/src/rg/transformators/expandGeneratorNodes.ts
import * as utils from '../../utils';
import * as ast from '../ast';

export function expandGeneratorNodes(gameDeclaration: ast.GameDeclaration) {
  for (const edge of gameDeclaration.edges.slice()) {
    if (ast.lib.hasBindings(edge.lhs) || ast.lib.hasBindings(edge.rhs)) {
      utils.remove(gameDeclaration.edges, edge);

      const bindings = [
        ...ast.lib.bindings(edge.lhs),
        ...ast.lib.bindings(edge.rhs),
      ].reduce<ast.Binding[]>(utils.unique, []);

      for (const mapping of createMappings(gameDeclaration, bindings)) {
        gameDeclaration.edges.push(
          ast.EdgeDeclaration({
            lhs: substituteBindings(edge.lhs, mapping),
            rhs: substituteBindings(edge.rhs, mapping),
            label: ast.lib.substituteInEdgeLabel(edge.label, mapping),
          }),
        );
      }
    }
  }
}

function createMappings(
  gameDeclaration: ast.GameDeclaration,
  bindings: ast.Binding[],
) {
  return bindings
    .map(binding =>
      ast.lib.typeValues(gameDeclaration, binding.type).map(ast.serializeValue),
    )
    .reduce<string[][]>(utils.cartesian, [[]])
    .map(values =>
      utils.mapToObject(bindings, (binding, index) => [
        binding.identifier,
        ast.Reference({ identifier: values[index] }),
      ]),
    );
}

function substituteBindings(
  edgeName: ast.EdgeName,
  mapping: Record<string, ast.Reference>,
) {
  if (!ast.lib.hasBindings) {
    return edgeName;
  }

  const identifier = edgeName.parts
    .map(part => {
      switch (part.kind) {
        case 'Binding':
          return mapping[part.identifier].identifier;
        case 'Literal':
          return part.identifier;
      }
    })
    .join('__bind__');

  return ast.EdgeName({ parts: [ast.Literal({ identifier })] });
}

File: interpreter_rust/rg_transform/src/expand_generator_nodes.rs
use rg::ast::{Error, GameDeclaration};
use std::rc::Rc;

pub fn expand_generator_nodes(
    mut game_declaration: GameDeclaration<String>,
) -> Result<GameDeclaration<String>, Error<String>> {
    game_declaration.edges = game_declaration
        .edges
        .iter()
        .map(|edge| {
            game_declaration
                .create_mappings(edge.bindings())
                .map(|mappings| {
                    mappings
                        .into_iter()
                        .map(|mapping| Rc::new(edge.substitute_bindings(&mapping)))
                })
        })
        .collect::<Result<Vec<_>, _>>()?
        .into_iter()
        .flatten()
        .collect();

    Ok(game_declaration)
}
